<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="lekiwi">

  <!-- Include robot properties -->
  <xacro:include filename="$(find lekiwi_base_control)/urdf/base_properties.xacro"/>

  <!-- Base link -->
  <link name="base_link">
    <visual>
      <geometry>
        <mesh filename="package://lekiwi_base_control/urdf/meshes/lekiwi_base.stl"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://lekiwi_base_control/urdf/meshes/lekiwi_base.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.02"/>
    </inertial>
  </link>

  <!-- Left wheel link -->
  <link name="left_wheel_link">
    <visual>
      <origin xyz="0 0 0" rpy="${pi} 0 0"/>
      <geometry>
        <mesh filename="package://lekiwi_base_control/urdf/meshes/omni_wheel.stl"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="${pi} 0 0"/>
      <geometry>
        <mesh filename="package://lekiwi_base_control/urdf/meshes/omni_wheel.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <!-- Back wheel link -->
  <link name="back_wheel_link">
    <visual>
      <origin xyz="0 0 0" rpy="${pi} 0 0"/>
      <geometry>
        <mesh filename="package://lekiwi_base_control/urdf/meshes/omni_wheel.stl"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="${pi} 0 0"/>
      <geometry>
        <mesh filename="package://lekiwi_base_control/urdf/meshes/omni_wheel.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <!-- Right wheel link -->
  <link name="right_wheel_link">
    <visual>
      <origin xyz="0 0 0" rpy="${pi} 0 0"/>
      <geometry>
        <mesh filename="package://lekiwi_base_control/urdf/meshes/omni_wheel.stl"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="${pi} 0 0"/>
      <geometry>
        <mesh filename="package://lekiwi_base_control/urdf/meshes/omni_wheel.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <!-- 60 degrees anti-clockwise from +X at radius base_radius -->
  <joint name="left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel_link"/>
    <origin xyz="${wheel_base_offset * cos(offset)} ${wheel_base_offset * sin(offset)} 0" rpy="0 ${pi/2} ${offset}"/>
    <axis xyz="0 0 1"/>
  </joint>

  <!-- 180 degrees anti-clockwise from +X at radius base_radius -->
  <joint name="back_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="back_wheel_link"/>
    <origin xyz="${wheel_base_offset * cos(offset + theta)} ${wheel_base_offset * sin(offset + theta)} 0" rpy="0 ${pi/2} ${offset + theta}"/>
    <axis xyz="0 0 1"/>
  </joint>

  <!-- 300 degrees (or -60 degrees) anti-clockwise from +X at radius base_radius -->
  <joint name="right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel_link"/>
    <origin xyz="${wheel_base_offset * cos(offset - theta)} ${wheel_base_offset * sin(offset - theta)} 0" rpy="0 ${pi/2} ${offset - theta}"/>
    <axis xyz="0 0 1"/>
  </joint>

  <!-- ROS2 Control System Interface -->
  <ros2_control name="lekiwi_base" type="system">
    <hardware>
      <plugin>sts_hardware_interface/STSHardwareInterface</plugin>

      <!-- Serial communication parameters -->
      <param name="serial_port">${serial_port}</param>
      <param name="baud_rate">${baud_rate}</param>
      <param name="use_sync_write">true</param>
      <param name="enable_multi_turn">true</param>
      <param name="enable_mock_mode">false</param>
    </hardware>

    <!-- Left wheel joint (Motor ID 7) -->
    <joint name="left_wheel_joint">
      <param name="motor_id">${left_motor_id}</param>
      <param name="operating_mode">${operating_mode}</param>
      <param name="max_velocity">${max_velocity_rads}</param>

      <command_interface name="velocity"/>
      <command_interface name="acceleration"/>

      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="load"/>
      <state_interface name="voltage"/>
      <state_interface name="temperature"/>
      <state_interface name="current"/>
      <state_interface name="is_moving"/>
    </joint>

    <!-- Back wheel joint (Motor ID 8) -->
    <joint name="back_wheel_joint">
      <param name="motor_id">${back_motor_id}</param>
      <param name="operating_mode">${operating_mode}</param>
      <param name="max_velocity">${max_velocity_rads}</param>

      <command_interface name="velocity"/>
      <command_interface name="acceleration"/>

      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="load"/>
      <state_interface name="voltage"/>
      <state_interface name="temperature"/>
      <state_interface name="current"/>
      <state_interface name="is_moving"/>
    </joint>

    <!-- Right wheel joint (Motor ID 9) -->
    <joint name="right_wheel_joint">
      <param name="motor_id">${right_motor_id}</param>
      <param name="operating_mode">${operating_mode}</param>
      <param name="max_velocity">${max_velocity_rads}</param>

      <command_interface name="velocity"/>
      <command_interface name="acceleration"/>

      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="load"/>
      <state_interface name="voltage"/>
      <state_interface name="temperature"/>
      <state_interface name="current"/>
      <state_interface name="is_moving"/>
    </joint>
  </ros2_control>

</robot>
